src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocPackageCheck.java

```diff
@@ -0,0 +1,159 @@
+ * >link</a> for more info.
+ * </p>
+ * <p>
+ * To configure the check:
+ * </p>
+ * <pre>
+ * &lt;module name="MissingJavadocPackage"/&gt;
+ * </pre>
+ * <p>Example:</p>
+ * <pre>
+ * /**
+ * * Provides API classes
+ * {@literal *}/
+ * package com.checkstyle.api; // no violation
+ * </pre>
+ *
+ * @since 8.22
+ */
+@StatelessCheck
+public class MissingJavadocPackageCheck extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_PKG_JAVADOC_MISSING = "package.javadoc.missing";
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {TokenTypes.PACKAGE_DEF};
+    }
+
+    @Override
+    public boolean isCommentNodesRequired() {
+        return true;
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        final FileContents contents = getFileContents();
+        if (contents.inPackageInfo() && !hasJavadocBefore(ast)) {
+            log(ast.getLineNo(), MSG_PKG_JAVADOC_MISSING);
+        }
+    }
+
+    /**
+     * Checks that there is javadoc before ast.
+     * Because of <a href="https://github.com/checkstyle/checkstyle/issues/4392">parser bug</a>
+     * parser can place javadoc comment either as previous sibling of package definition, e.g.<br/>
+     * <pre>
+     * BLOCK_COMMENT_BEGIN -> /* [1:0]
+     * |--COMMENT_CONTENT -> *\r\n* Important javadoc\r\n [1:2]
+     * `--BLOCK_COMMENT_END ->  {@literal *}/ [3:0]
+     * PACKAGE_DEF ->package[4:0]
+     * |--ANNOTATIONS ->ANNOTATIONS [4:17]
+     * |--DOT ->.[4:17]
+     * | |--DOT ->.[4:11]
+     * | | |--IDENT ->com [4:8]
+     * | | `--IDENT ->mypack [4:12]
+     * | `--IDENT ->checkstyle [4:18]
+     * `--SEMI -> ; [4:28]
+     * </pre>
+     * or (if there is annotation between package def and javadoc)<br/>
+     * inside package definition tree, e.g.
+     * <pre>
+     * PACKAGE_DEF -> package [5:0]
+     * |--ANNOTATIONS -> ANNOTATIONS [4:0]
+     * |   `--ANNOTATION -> ANNOTATION [4:0]
+     * |       |--BLOCK_COMMENT_BEGIN -> /* [1:0]
+     * |       |   |--COMMENT_CONTENT -> *\r\n* Important javadoc\r\n [1:2]
+     * |       |   `--BLOCK_COMMENT_END -> {@literal *}/ [3:0]
+     * |       |--AT -> @ [4:0]
+     * |`--IDENT -> Deprecated [4:1]
+     * |--DOT -> . [5:17]
+     * |   |--DOT -> . [5:11]
+     * |   |   |--IDENT -> com [5:8]
+     * |   |   `--IDENT -> mypack [5:12]
+     * |   `--IDENT -> checkstyle [5:18]
+     * `--SEMI ->; [5:28]
+     * </pre>
+     * So we should look for javadoc in both places.
+     *
+     * @param ast {@link TokenTypes#PACKAGE_DEF} token to check
+     * @return true if there is javadoc, false otherwise
+     */
+    private static boolean hasJavadocBefore(DetailAST ast) {
```

Comment:
javadoc is always "before" for engineer in java code, before he get to a point of nuances of our AST.
Please name it `hasJavadoc`, to hide fact that implementation will search in both places.

Please review JavadocUtil methods, there should be something useful for you to reuse. `hasJavadoc` is kind of very common method that most javadoc checks that come from java AST to comments AST should use it.

Reply:
done